File: sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        parent = [-1] * len(nums)
        def find(x):
            if parent[x] == -1:
                return x
            parent[x] = find(parent[x])
            return parent[x]
        stack = []
        for i, el in enumerate(nums):
            while stack and el > nums[stack[-1]]:
                curr = stack.pop()
                parent[curr] = find(i)
            stack.append(i)
            if i >= k - 1:
                res.append(nums[find(i - k + 1)])
        return res

File: sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        deque<int> dq;
        for (int i = 0; i < nums.size(); i++) {
            if (!dq.empty() && dq.front() == i - k) dq.pop_front();
            while (!dq.empty() && nums[dq.back()] < nums[i]) dq.pop_back();
            dq.push_back(i);
            if (i >= k - 1) result.push_back(nums[dq.front()]);
        }
        return result;
    }
};